airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_incremental.py

```diff
@@ -170,70 +180,49 @@ async def test_read_sequential_slices(
 
-            # if last state message, skip
-            # this is because we cannot assert if the last state message will result in new records
-            # as in this case it is possible for a connector to return a previous state message.
-            # e.g. if the connector is using pagination and the last page is only partially full
-            if idx == len(unique_state_messages) - 1:
-                continue
+            states_with_expected_record_count = self._state_messages_selector(states_1)
+            mutating_stream_name_to_per_stream_state = dict()
 
-            # if batching required, and not a sample, skip
-            if len(unique_state_messages) >= min_batches_to_test and idx % sample_rate != 0:
-                continue
+            for idx, state_message_data in enumerate(states_with_expected_record_count):
+                state_message, expected_records_count = state_message_data
+                assert state_message.type == Type.STATE
 
-            state_input, mutating_stream_name_to_per_stream_state = self.get_next_state_input(
-                state_message, mutating_stream_name_to_per_stream_state, is_per_stream
-            )
+                state_input, mutating_stream_name_to_per_stream_state = self.get_next_state_input(
+                    state_message, mutating_stream_name_to_per_stream_state
+                )
 
-            output_N = await docker_runner.call_read_with_state(connector_config, configured_catalog_for_incremental, state=state_input)
-            records_N = filter_output(output_N, type_=Type.RECORD)
-            assert (
-                records_N
-            ), f"Read {idx + 2} of {len(unique_state_messages)} should produce at least one record.\n\n state: {state_input} \n\n records_{idx + 2}: {records_N}"
+                output_N = await docker_runner.call_read_with_state(
+                    connector_config, configured_catalog_for_incremental_per_stream, state=state_input
+                )
+                records_N = filter_output(output_N, type_=Type.RECORD)
+
+                assert (
+                    # We assume that the output may be empty when we read the latest state, or it must produce some data if we are in the middle of our progression
+                    len(records_N) >= expected_records_count
+                    or idx == len(states_with_expected_record_count) - 1
+                ), f"Read {idx + 1} of {len(states_with_expected_record_count)} should produce at least one record.\n\n state: {state_input} \n\n records_{idx + 1}: {records_N}"
 
-            diff = naive_diff_records(records_1, records_N)
-            assert (
-                diff
-            ), f"Records for subsequent reads with new state should be different.\n\n records_1: {records_1} \n\n state: {state_input} \n\n records_{idx + 2}: {records_N} \n\n diff: {diff}"
+                diff = naive_diff_records(records_1, records_N)
+                assert (
+                    diff
```

Comment:
Do we need to check for diff here? An example of correct read results where the record filter is inclusive from both sides, for which tests fail:
```
[
    {"type": "STATE", "name": "test_stream", "stream_state": {}, "sourceStats": {"recordCount": 0.0}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-08"}},
    {"type": "STATE", "name": "test_stream", "stream_state": {"date": "2022-05-08"}, "sourceStats": {"recordCount": 1.0}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-08"}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-09"}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-10"}},
    {"type": "STATE", "name": "test_stream", "stream_state": {"date": "2022-05-10"}, "sourceStats": {"recordCount": 2.0}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-10"}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-11"}},
    {"type": "RECORD", "name": "test_stream", "data": {"date": "2022-05-12"}},
    {"type": "STATE", "name": "test_stream", "stream_state": {"date": "2022-05-12"}, "sourceStats": {"recordCount": 3.0}}
]
```

Reply:
I think we need it because we can't rely only on record count

Reply:
can you add a comment documenting this rationale @roman-yermilov-gl ?

Reply:
I created the issue for further investigation: https://github.com/airbytehq/airbyte-internal-issues/issues/8056

Reply:
@tolik0 is going to prepare a new ticket to investigate this rationale. In the scope of this ticket, he will also do measurements and possibly make improvements for connectors that run slowly due to the diff operation